internal/directives/http_requester.go

```diff
@@ -0,0 +1,253 @@
+	r := &httpRequester{}
+	r.schemaLoader = getConfigSchemaLoader(r.Name())
+	return r
+}
+
+// Name implements the PromotionStepRunner interface.
+func (h *httpRequester) Name() string {
+	return "http"
+}
+
+func (h *httpRequester) RunPromotionStep(
+	ctx context.Context,
+	stepCtx *PromotionStepContext,
+) (PromotionStepResult, error) {
+	if err := h.validate(stepCtx.Config); err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored}, err
+	}
+	cfg, err := ConfigToStruct[HTTPConfig](stepCtx.Config)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("could not convert config into http config: %w", err)
+	}
+	return h.runPromotionStep(ctx, stepCtx, cfg)
+}
+
+// validate validates httpRequester configuration against a JSON schema.
+func (h *httpRequester) validate(cfg Config) error {
+	return validate(h.schemaLoader, gojsonschema.NewGoLoader(cfg), h.Name())
+}
+
+func (h *httpRequester) runPromotionStep(
+	_ context.Context,
+	_ *PromotionStepContext,
+	cfg HTTPConfig,
+) (PromotionStepResult, error) {
+	req, err := h.buildRequest(cfg)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error building HTTP request: %w", err)
+	}
+	resp, err := h.getClient(cfg).Do(req)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error sending HTTP request: %w", err)
+	}
+	defer resp.Body.Close()
+	env, err := h.buildExprEnv(resp)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error building expression context from HTTP response: %w", err)
+	}
+	success, err := h.wasRequestSuccessful(cfg, resp.StatusCode, env)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error evaluating success criteria: %w", err)
+	}
+	failure, err := h.didRequestFail(cfg, resp.StatusCode, env)
+	if err != nil {
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+			fmt.Errorf("error evaluating failure criteria: %w", err)
+	}
+	switch {
+	case success && !failure:
+		outputs, err := h.buildOutputs(cfg.Outputs, env)
+		if err != nil {
+			return PromotionStepResult{Status: kargoapi.PromotionPhaseErrored},
+				fmt.Errorf("error extracting outputs from HTTP response: %w", err)
+		}
+		return PromotionStepResult{
+			Status: kargoapi.PromotionPhaseSucceeded,
+			Output: outputs,
+		}, nil
+	case failure:
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseFailed},
+			fmt.Errorf("HTTP response met failure criteria")
+	default:
+		return PromotionStepResult{Status: kargoapi.PromotionPhaseRunning}, nil
+	}
+}
+
+func (h *httpRequester) buildRequest(cfg HTTPConfig) (*http.Request, error) {
+	method := cfg.Method
+	if method == "" {
+		method = http.MethodGet
+	}
+	req, err := http.NewRequest(method, cfg.URL, bytes.NewBufferString(cfg.Body))
```

Comment:
We should probably set a max size for the buffer and/or check the size of the response.

Reply:
What maximum did you have in mind? I'm thinking a 2k would be very generous.

Reply:
I would set it to something like 2MB to allow almost any response, as long as they are not too outrageous for the memory consumption.